In [1]:
# pip install plotly

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import plotly.express as px

# Initial EDA

In [3]:
# # Mount GDrive folders
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# Set home directory
import os
home_dir = "drive/MyDrive/Eskwelabs/Sprint 5 - Capstone/data"
os.chdir(home_dir)

In [5]:
df = pd.read_csv('DOE- Data Request- Sample Daily Data_GEN_2023924.csv')

<ipython-input-5-d6f05d41c83f>:1: DtypeWarning: Columns (9,14,66,69) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('DOE- Data Request- Sample Daily Data_GEN_2023924.csv')


In [6]:
# Load the data
# df = pd.read_csv('E:\Documents\Google Drive\Eskwelabs\Sprint 5 - Capstone\data\DOE- Data Request- Sample Daily Data_GEN_2023924.csv')

In [7]:
pd.set_option('display.max_columns', None)
df.head()

time_interval    delivery_date  delivery_hour    resource_id  \
0  24/09/2023 0:05  24/09/2023 0:00              1  01BTSOLEN_BAT   
1  24/09/2023 0:05  24/09/2023 0:00              1    01CNCEP_BAT   
2  24/09/2023 0:05  24/09/2023 0:00              1    01LAMAO_BAT   
3  24/09/2023 0:05  24/09/2023 0:00              1    01LIMAY_BAT   
4  24/09/2023 0:05  24/09/2023 0:00              1  01MAGAPIT_BAT   

        resource_name trading_id major_part_id_new type_id bid_type nsch_type  \
0      BATAAN BATTERY       BSEI                AC     GEN      REG       NaN   
1  CONCEPCION BATTERY       UPSI               SMC     GEN      REG       NaN   
2       LAMAO BATTERY       UPSI               SMC     GEN      REG       NaN   
3       LIMAY BATTERY       UPSI               SMC     GEN      REG       NaN   
4     MAGAPIT BATTERY       UPSI               SMC     GEN      REG       NaN   

  plant_type region_id zone_id  max_cap_rtd Outage_Tag  outage_capacity_rtd  \
0        BAT      CLUZ  NLUZON          0.5        NaN                  NaN   
1        BAT      CLUZ  NLUZON         60.0        NaN                  NaN   
2        BAT      CLUZ  NLUZON         50.0        NaN                  NaN   
3        BAT      CLUZ  NLUZON         40.0        NaN                  NaN   
4        BAT      CLUZ  NLUZON         40.0        NaN                  NaN   

   rtd_price1  rtd_quantity1  rtd_price2  rtd_quantity2  rtd_price3  \
0         NaN            NaN         NaN            NaN         NaN   
1         NaN            NaN         NaN            NaN         NaN   
2    -10000.0          -50.0    -10000.0            0.0     32000.0   
3    -10000.0          -40.0    -10000.0            0.0     32000.0   
4         NaN            NaN         NaN            NaN         NaN   

   rtd_quantity3  rtd_price4  rtd_quantity4  rtd_price5  rtd_quantity5  \
0            NaN         NaN            NaN         NaN            NaN   
1            NaN         NaN            NaN         NaN            NaN   
2           50.0         NaN            NaN         NaN            NaN   
3           40.0         NaN            NaN         NaN            NaN   
4            NaN         NaN            NaN         NaN            NaN   

   rtd_price6  rtd_quantity6  rtd_price7  rtd_quantity7  rtd_price8  \
0         NaN            NaN         NaN            NaN         NaN   
1         NaN            NaN         NaN            NaN         NaN   
2         NaN            NaN         NaN            NaN         NaN   
3         NaN            NaN         NaN            NaN         NaN   
4         NaN            NaN         NaN            NaN         NaN   

   rtd_quantity8  rtd_price9  rtd_quantity9  rtd_price10  rtd_quantity10  \
0            NaN         NaN            NaN          NaN             NaN   
1            NaN         NaN            NaN          NaN             NaN   
2            NaN         NaN            NaN          NaN             NaN   
3            NaN         NaN            NaN          NaN             NaN   
4            NaN         NaN            NaN          NaN             NaN   

   rtd_price11  rtd_quantity11  rtd_rr_break_quantity1  rtd_rr_up1  \
0          NaN             NaN                     NaN         NaN   
1          NaN             NaN                     NaN         NaN   
2          NaN             NaN                    50.0      2300.4   
3          NaN             NaN                    40.0      2642.4   
4          NaN             NaN                     NaN         NaN   

   rtd_rr_down1  rtd_rr_break_quantity2  rtd_rr_up2  rtd_rr_down2  \
0           NaN                     NaN         NaN           NaN   
1           NaN                     NaN         NaN           NaN   
2        2300.4                     NaN         NaN           NaN   
3        2642.4                     NaN         NaN           NaN   
4           NaN                     NaN         NaN           NaN   

   rtd_rr_break_quantity3  rtd_rr_up

In [8]:
df.tail()

time_interval    delivery_date  delivery_hour   resource_id  \
120667  24/09/2023 23:50  24/09/2023 0:00             24  08SLWIND_G01   
120668  24/09/2023 23:55  24/09/2023 0:00             24   08PWIND_G01   
120669  24/09/2023 23:55  24/09/2023 0:00             24  08SLWIND_G01   
120670   25/09/2023 0:00  24/09/2023 0:00             24   08PWIND_G01   
120671   25/09/2023 0:00  24/09/2023 0:00             24  08SLWIND_G01   

           resource_name trading_id major_part_id_new type_id bid_type  \
120667  SAN LORENZO WIND      TAREC                AC     GEN     NSCH   
120668   PWEI NABAS WIND       PWEI             BWCUA     GEN     NSCH   
120669  SAN LORENZO WIND      TAREC                AC     GEN     NSCH   
120670   PWEI NABAS WIND       PWEI             BWCUA     GEN     NSCH   
120671  SAN LORENZO WIND      TAREC                AC     GEN     NSCH   

       nsch_type plant_type region_id zone_id  max_cap_rtd Outage_Tag  \
120667      MUST       WIND      CVIS   PANAY         54.0        NaN   
120668      MUST       WIND      CVIS   PANAY         36.7        NaN   
120669      MUST       WIND      CVIS   PANAY         54.0        NaN   
120670      MUST       WIND      CVIS   PANAY         36.7        NaN   
120671      MUST       WIND      CVIS   PANAY         54.0        NaN   

        outage_capacity_rtd  rtd_price1  rtd_quantity1  rtd_price2  \
120667                  NaN         NaN            NaN         NaN   
120668                  NaN         NaN            NaN         NaN   
120669                  NaN         NaN            NaN         NaN   
120670                  NaN         NaN            NaN         NaN   
120671                  NaN         NaN            NaN         NaN   

        rtd_quantity2  rtd_price3  rtd_quantity3  rtd_price4  rtd_quantity4  \
120667            NaN         NaN            NaN         NaN            NaN   
120668            NaN         NaN            NaN         NaN            NaN   
120669            NaN         NaN            NaN         NaN            NaN   
120670            NaN         NaN            NaN         NaN            NaN   
120671            NaN         NaN            NaN         NaN            NaN   

        rtd_price5  rtd_quantity5  rtd_price6  rtd_quantity6  rtd_price7  \
120667         NaN            NaN         NaN            NaN         NaN   
120668         NaN            NaN         NaN            NaN         NaN   
120669         NaN            NaN         NaN            NaN         NaN   
120670         NaN            NaN         NaN            NaN         NaN   
120671         NaN            NaN         NaN            NaN         NaN   

        rtd_quantity7  rtd_price8  rtd_quantity8  rtd_price9  rtd_quantity9  \
120667            NaN         NaN            NaN         NaN            NaN   
120668            NaN         NaN            NaN         NaN            NaN   
120669            NaN         NaN            NaN         NaN            NaN   
120670            NaN         NaN            NaN         NaN            NaN   
120671            NaN         NaN            NaN         NaN            NaN   

        rtd_price10  rtd_quantity10  rtd_price11  rtd_quantity11  \
120667          NaN             NaN          NaN             NaN   
120668          NaN             NaN          NaN             NaN   
120669          NaN             NaN          NaN             NaN   
120670          NaN             NaN          NaN             NaN   
120671          NaN             NaN          NaN             NaN   

        rtd_rr_break_quantity1  rtd_rr_up1  rtd_rr_down1  \
120667                     NaN         NaN           NaN   
120668                     NaN         NaN           NaN   
120669                     NaN         NaN           NaN   
120670                     NaN         NaN           NaN   
120671                     NaN         NaN           NaN   

        rtd_rr_break_quantity2  rtd_rr_up2  rtd_rr_down2  \
120667                  

In [9]:
df_shape = df.shape
df_shape

(120672, 108)

Columns 9, 14, 66, and 69 have mixed dtypes:

9    nsch_type                   object

14   Outage_Tag                  object

66   com_status                  object

69   marginal_plant              object


In [10]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120672 entries, 0 to 120671
Data columns (total 108 columns):
 #    Column                      Dtype  
---   ------                      -----  
 0    time_interval               object 
 1    delivery_date               object 
 2    delivery_hour               int64  
 3    resource_id                 object 
 4    resource_name               object 
 5    trading_id                  object 
 6    major_part_id_new           object 
 7    type_id                     object 
 8    bid_type                    object 
 9    nsch_type                   object 
 10   plant_type                  object 
 11   region_id                   object 
 12   zone_id                     object 
 13   max_cap_rtd                 float64
 14   Outage_Tag                  object 
 15   outage_capacity_rtd         float64
 16   rtd_price1                  float64
 17   rtd_quantity1               float64
 18   rtd_price2                  float64
 19   

## Fix 2 datatypes in 1 column error

In [11]:
#determine why they contain unique values
unique_9 = pd.unique(df['nsch_type'])
unique_9

array([nan, 'NSCH', 'PRI', 'MUST'], dtype=object)

In [12]:
unique_14 = pd.unique(df['Outage_Tag'])
unique_14

array([nan, 'OUT'], dtype=object)

In [13]:
unique_66 = pd.unique(df['com_status'])
unique_66

array(['NOT', nan], dtype=object)

In [14]:
unique_69 = pd.unique(df['marginal_plant'])
unique_69

array([nan, 'Y'], dtype=object)

In [15]:
#transform nan datatype into empty string
df['nsch_type'] = df['nsch_type'].fillna('').astype(str)
df['Outage_Tag'] = df['Outage_Tag'].fillna('').astype(str)
df['com_status'] = df['com_status'].fillna('').astype(str)
df['marginal_plant'] = df['marginal_plant'].fillna('').astype(str)

## Check outliers

In [16]:
#check for outliers in the data
df.describe()

delivery_hour    max_cap_rtd  outage_capacity_rtd    rtd_price1  \
count  120672.000000  120672.000000          7582.000000  62197.000000   
mean       12.500000      63.253461           132.336956  12154.107862   
std         6.922215     106.245409           146.474776  18308.400245   
min         1.000000       0.300000             2.000000 -10000.000000   
25%         6.750000       7.800000            30.000000 -10000.000000   
50%        12.500000      20.000000           116.000000  12000.000000   
75%        18.250000      70.000000           151.400000  32000.000000   
max        24.000000     668.000000           668.000000  32000.000000   

       rtd_quantity1    rtd_price2  rtd_quantity2    rtd_price3  \
count   62197.000000  62197.000000   62197.000000  29164.000000   
mean       -0.783896  12154.107862      42.750840   9117.087533   
std         5.439502  18308.400245      64.896571  13478.687784   
min       -50.000000 -10000.000000       0.000000  -9999.000000   
25%         0.000000 -10000.000000       6.500000      0.000000   
50%         0.000000  12000.000000      15.100000   3330.000000   
75%         0.000000  32000.000000      50.000000  24700.000000   
max         0.000000  32000.000000     575.000000  32000.000000   

       rtd_quantity3    rtd_price4  rtd_quantity4    rtd_price5  \
count   29164.000000  15492.000000   15492.000000   6941.000000   
mean       99.871571  13425.123170     111.314246   6052.017405   
std       117.942569  12904.295519      93.335277   7434.987626   
min         2.000000  -9998.000000       3.000000  -9997.000000   
25%        26.500000   4150.080000      55.000000   3070.000000   
50%        50.000000   6000.090000      83.700000   4512.400000   
75%       138.000000  25000.000000     120.000000   5896.000000   
max       668.000000  32000.000000     408.600000  32000.000000   

       rtd_quantity5    rtd_price6  rtd_quantity6    rtd_price7  \
count    6941.000000   4259.000000    4259.000000   3556.000000   
mean      188.150944   7318.447222     200.311458   7335.534443   
std       121.084155   8172.646403     135.459002   7463.564602   
min         4.000000   2177.000000       6.000000   2180.000000   
25%       100.000000   3183.000000     110.000000   5000.000000   
50%       150.000000   5040.900000     120.000000   5759.300000   
75%       316.000000   6500.080000     280.000000   6904.000000   
max       400.000000  32000.000000     450.000000  31999.000000   

       rtd_quantity7    rtd_price8  rtd_quantity8   rtd_price9  rtd_quantity9  \
count    3556.000000   3460.000000    3460.000000  2138.000000    2138.000000   
mean      235.810461   8349.961237     260.097341  5532.131469     303.779701   
std       152.499839   8163.005846     167.964665  1928.269498     201.158616   
min         7.000000   2204.000000       8.000000  2238.000000      10.000000   
25%       120.000000   5512.600000     130.000000  3155.000000     140.000000   
50%       135.000000   6000.000000     151.300000  5900.000000     315.000000   
75%       300.000000   7400.080000     330.000000  7800.080000     600.000000   
max       500.000000  32000.000000     550.000000  7950.080000     600.000000   

       rtd_price10  rtd_quantity10  rtd_price11  rtd_quantity11  \
count  1152.000000     1152.000000          0.0             0.0   
mean   5375.290000      391.725000          NaN             NaN   
std    2646.673628      242.068466          NaN             NaN   
min    2283.000000      149.900000          NaN             NaN   
25%    2984.250000      149.975000          NaN             NaN   
50%    5609.040000      385.000000          NaN             NaN   
75%    8000.080000      626.750000          NaN             NaN   
max    8000.080000      647.000000          NaN             NaN   

       rtd_rr_break_quantity1    rtd_rr_up1  rtd_rr_down1  \
count            62197.000000  62197.000000  62197.000000   
mean                55.892151     93.081316     91.973245   
std        

In [17]:
# Check if 5-minute timestamps equal to 288 for this 24-hour dataset
timestamp_counts = df['time_interval'].value_counts()
print(timestamp_counts)

# Count the number of unique timestamps
num_unique_timestamps = len(df['time_interval'].unique())
print(f"\nNumber of unique timestamps: {num_unique_timestamps}")

24/09/2023 0:05     419
24/09/2023 0:10     419
24/09/2023 16:25    419
24/09/2023 16:20    419
24/09/2023 16:15    419
                   ... 
24/09/2023 8:00     419
24/09/2023 7:55     419
24/09/2023 7:50     419
24/09/2023 7:45     419
25/09/2023 0:00     419
Name: time_interval, Length: 288, dtype: int64

Number of unique timestamps: 288


In [18]:
df

time_interval    delivery_date  delivery_hour    resource_id  \
0        24/09/2023 0:05  24/09/2023 0:00              1  01BTSOLEN_BAT   
1        24/09/2023 0:05  24/09/2023 0:00              1    01CNCEP_BAT   
2        24/09/2023 0:05  24/09/2023 0:00              1    01LAMAO_BAT   
3        24/09/2023 0:05  24/09/2023 0:00              1    01LIMAY_BAT   
4        24/09/2023 0:05  24/09/2023 0:00              1  01MAGAPIT_BAT   
...                  ...              ...            ...            ...   
120667  24/09/2023 23:50  24/09/2023 0:00             24   08SLWIND_G01   
120668  24/09/2023 23:55  24/09/2023 0:00             24    08PWIND_G01   
120669  24/09/2023 23:55  24/09/2023 0:00             24   08SLWIND_G01   
120670   25/09/2023 0:00  24/09/2023 0:00             24    08PWIND_G01   
120671   25/09/2023 0:00  24/09/2023 0:00             24   08SLWIND_G01   

             resource_name trading_id major_part_id_new type_id bid_type  \
0           BATAAN BATTERY       BSEI                AC     GEN      REG   
1       CONCEPCION BATTERY       UPSI               SMC     GEN      REG   
2            LAMAO BATTERY       UPSI               SMC     GEN      REG   
3            LIMAY BATTERY       UPSI               SMC     GEN      REG   
4          MAGAPIT BATTERY       UPSI               SMC     GEN      REG   
...                    ...        ...               ...     ...      ...   
120667    SAN LORENZO WIND      TAREC                AC     GEN     NSCH   
120668     PWEI NABAS WIND       PWEI             BWCUA     GEN     NSCH   
120669    SAN LORENZO WIND      TAREC                AC     GEN     NSCH   
120670     PWEI NABAS WIND       PWEI             BWCUA     GEN     NSCH   
120671    SAN LORENZO WIND      TAREC                AC     GEN     NSCH   

       nsch_type plant_type region_id zone_id  max_cap_rtd Outage_Tag  \
0                       BAT      CLUZ  NLUZON          0.5              
1                       BAT      CLUZ  NLUZON         60.0              
2                       BAT      CLUZ  NLUZON         50.0              
3                       BAT      CLUZ  NLUZON         40.0              
4                       BAT      CLUZ  NLUZON         40.0              
...          ...        ...       ...     ...          ...        ...   
120667      MUST       WIND      CVIS   PANAY         54.0              
120668      MUST       WIND      CVIS   PANAY         36.7              
120669      MUST       WIND      CVIS   PANAY         54.0              
120670      MUST       WIND      CVIS   PANAY         36.7              
120671      MUST       WIND      CVIS   PANAY         54.0              

        outage_capacity_rtd  rtd_price1  rtd_quantity1  rtd_price2  \
0                       NaN         NaN            NaN         NaN   
1                       NaN         NaN            NaN         NaN   
2                       NaN    -10000.0          -50.0    -10000.0   
3                       NaN    -10000.0          -40.0    -10000.0   
4                       NaN         NaN            NaN         NaN   
...                     ...         ...            ...         ...   
120667                  NaN         NaN            NaN         NaN   
120668                  NaN         NaN            NaN         NaN   
120669                  NaN         NaN            NaN         NaN   
120670                  NaN         NaN            NaN         NaN   
120671                  NaN         NaN            NaN         NaN   

        rtd_quantity2  rtd_price3  rtd_quantity3  rtd_price4  rtd_quantity4  \
0                 NaN         NaN            NaN         NaN            NaN   
1                 NaN         NaN            NaN         NaN            NaN   
2                 0.0     32000.0           50.0         NaN            NaN   
3                 0.0     32000.0           40.0         NaN            NaN   
4                 NaN         NaN            NaN         NaN            NaN   
...  

In [19]:
# Assuming your DataFrame is named df
resource_id_counts = df['resource_id'].value_counts()

# Set display options to show all rows and columns
pd.set_option('display.max_rows', None)

# Display the unique names and their counts
resource_id_counts

01BTSOLEN_BAT     288
06HPCO_G01        288
11PACERM_G01      288
11NBPC_G01        288
11MNRGY_G02       288
11MNRGY_G01       288
11KEGMAR_G01      288
10MEGC_G01        288
10IDPP_G01        288
09WMPC_U10        288
09WMPC_U09        288
09WMPC_U08        288
09WMPC_U07        288
09WMPC_U06        288
09WMPC_U05        288
09WMPC_U04        288
09WMPC_U03        288
09WMPC_U02        288
09WMPC_U01        288
09PANABNK_U02     288
09PANABNK_U01     288
09KEGJIM_G01      288
14MARBEL_U01      288
14ALMADA_G01      288
13TUDAY2_G01      288
11PKBUK_G01       288
12KEGGIN_G01      288
12KEGTAN_G01      288
11KBSOL_G01       288
06CABI_G01        288
06BISCOM_G01      288
07UBAY_BAT        288
06KABAN_BAT       288
05TOLEDO_BAT      288
04ORMOC_BAT       288
14NVSOL_G01       288
14ASTROSOL_G01    288
13DIGSOL_G01      288
11KIRAS_G01       288
14SUPKOR_G01      288
12NACSUR_G01      288
14PKPSOC_G01      288
13TM1_U02         288
13TM1_U01         288
13MPCDIG_G01      288
13MATIBNK_

In [20]:
# Reset display options to default
pd.set_option('display.max_rows', 10)
# pd.set_option('display.max_columns', 20)

In [21]:
num_unique_resource_ids = df['resource_id'].nunique()
num_unique_resource_ids

419


There are 419 participants, with each having an entry for a timestamp.
419 * 288 5-min intervals in 24h = 120,672 total rows.

## Convert to date-time

In [22]:
# Convert 'time_interval' to datetime format
df['time_interval'] = pd.to_datetime(df['time_interval'], format='%d/%m/%Y %H:%M')

# Participating Unit Information

In [44]:
# how many participants in the df
major_id_counts = df['major_part_id_new'].value_counts()
major_comp_units = major_id_counts / 288
major_comp_units

AP       71.0
SMC      34.0
AC       29.0
PSALM    29.0
FGC      26.0
         ... 
SMTI      1.0
RSEHI     1.0
MGEN      1.0
FCVC      1.0
BWCUA     1.0
Name: major_part_id_new, Length: 123, dtype: float64

In [24]:
# Major companies and their generating companies
result = df.groupby('major_part_id_new')['resource_id'].unique().reset_index()
result

major_part_id_new                                        resource_id
0                  AC  [01BTSOLEN_BAT, 03ALMNOS_BAT1, 03ALMNOS_BAT2, ...
1               ACNPC                                      [01ACNPC_G01]
2                  AE  [09WMPC_U01, 09WMPC_U02, 09WMPC_U03, 09WMPC_U0...
3                AEHC                                      [01YHGRN_G01]
4                 AEL                                       [01ANDA_G01]
..                ...                                                ...
118         TRUSTSOLR                                   [01TRUSTSOL_G01]
119               UEC  [01CLEANG_G01, 01GRGOLD_G01, 01IPOWER_G01, 01I...
120               VEC         [01BAUANG_GS1, 01BAUANG_GS2, 01BAUANG_GS3]
121               VMC                             [06VMC_G01, 06VMC_G02]
122              VSEC                                     [02VALSOL_G01]

[123 rows x 2 columns]

In [41]:
# Assuming you have a DataFrame named 'result' with a column 'resource_id' containing lists
result['resource_id_count'] = result['resource_id'].apply(lambda x: len(x) if isinstance(x, list) else 0)

# Display the DataFrame with the new 'resource_id_count' column
result


major_part_id_new                                        resource_id  \
0                  AC  [01BTSOLEN_BAT, 03ALMNOS_BAT1, 03ALMNOS_BAT2, ...   
1               ACNPC                                      [01ACNPC_G01]   
2                  AE  [09WMPC_U01, 09WMPC_U02, 09WMPC_U03, 09WMPC_U0...   
3                AEHC                                      [01YHGRN_G01]   
4                 AEL                                       [01ANDA_G01]   
..                ...                                                ...   
118         TRUSTSOLR                                   [01TRUSTSOL_G01]   
119               UEC  [01CLEANG_G01, 01GRGOLD_G01, 01IPOWER_G01, 01I...   
120               VEC         [01BAUANG_GS1, 01BAUANG_GS2, 01BAUANG_GS3]   
121               VMC                             [06VMC_G01, 06VMC_G02]   
122              VSEC                                     [02VALSOL_G01]   

     resource_id_count  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
..                 ...  
118                  0  
119                  0  
120                  0  
121                  0  
122                  0  

[123 rows x 3 columns]

In [25]:
# Make a dict for resource_id
resource_dict = {}

for resource_id, group in df.groupby('resource_id'):
    resource_dict[resource_id] = {
        'resource_name': group['resource_name'].unique().tolist(),
        'major_part_id_new': group['major_part_id_new'].unique().tolist(),
        'plant_type': group['plant_type'].unique().tolist(),
        'max_cap_rtd': group['max_cap_rtd'].unique().tolist(),
        'region_id': group['region_id'].unique().tolist(),
        'zone_id': group['zone_id'].unique().tolist(),
    }

# Display the dictionary
print(resource_dict)

{'01ACNPC_G01': {'resource_name': ['ACNPC BIOMASS'], 'major_part_id_new': ['ACNPC'], 'plant_type': ['BIOF'], 'max_cap_rtd': [3.2], 'region_id': ['CLUZ'], 'zone_id': ['NLUZON']}, '01AMBUK_U01': {'resource_name': ['AMBUKLAO HEP'], 'major_part_id_new': ['AP'], 'plant_type': ['HYD'], 'max_cap_rtd': [37.5], 'region_id': ['CLUZ'], 'zone_id': ['NLUZON']}, '01AMBUK_U02': {'resource_name': ['AMBUKLAO HEP'], 'major_part_id_new': ['AP'], 'plant_type': ['HYD'], 'max_cap_rtd': [37.5], 'region_id': ['CLUZ'], 'zone_id': ['NLUZON']}, '01AMBUK_U03': {'resource_name': ['AMBUKLAO HEP'], 'major_part_id_new': ['AP'], 'plant_type': ['HYD'], 'max_cap_rtd': [37.5], 'region_id': ['CLUZ'], 'zone_id': ['NLUZON']}, '01AMPHAW_G01': {'resource_name': ['AMPOHAW HEP'], 'major_part_id_new': ['AP'], 'plant_type': ['HYD'], 'max_cap_rtd': [12.5], 'region_id': ['CLUZ'], 'zone_id': ['NLUZON']}, '01ANDA_G01': {'resource_name': ['ANDA CFTPP'], 'major_part_id_new': ['AEL'], 'plant_type': ['COAL'], 'max_cap_rtd': [72.0], 'regi

In [26]:
# Function to print values for a given resource_name
def print_resource_info(resource_name):
    if resource_name in resource_dict:
        values = resource_dict[resource_name]
        print(f"Resource Name: {values['resource_name']}")
        print(f"Company: {values['major_part_id_new']}")
        print(f"Plant Type: {values['plant_type']}")
        print(f"Max Capacity (MW): {values['max_cap_rtd']}")
        print(f"Region: {values['region_id']}")
        print(f"Zone: {values['zone_id']}")
    else:
        print(f"Resource with name '{resource_name}' not found.")

In [27]:
# Input resource name to get its information
input_resource_name = '01AMBUK_U01'#'01SUAL_G01'
print_resource_info(input_resource_name)

Resource Name: ['AMBUKLAO HEP']
Company: ['AP']
Plant Type: ['HYD']
Max Capacity (MW): [37.5]
Region: ['CLUZ']
Zone: ['NLUZON']


# Major Players

In [28]:
# Convert the dictionary to a list of dictionaries for Plotly
data_for_plotly = []
for resource_id, values in resource_dict.items():
    total_max_cap_rtd = sum(values['max_cap_rtd'])  # Calculate total max capacity for the resource_id
    for plant_type, resource_name in zip(values['plant_type'], values['resource_name']):
        data_for_plotly.append({
            'major_part_id_new': values['major_part_id_new'][0],  # Assuming major_part_id_new is the same for all entries
            'plant_type': plant_type,
            'total_max_cap_rtd': total_max_cap_rtd,
            'resource_name': resource_name,
            'resource_id': resource_id
        })

# Create DataFrame from the list of dictionaries
df_for_plotly = pd.DataFrame(data_for_plotly)

# Sort the DataFrame by total_max_cap_rtd in descending order
df_for_plotly = df_for_plotly.sort_values(by='total_max_cap_rtd', ascending=False)

# # Create a stacked bar chart using Plotly Express
# fig = px.bar(df_for_plotly,
#              x='major_part_id_new',
#              y='total_max_cap_rtd',
#              color='plant_type',
#              labels={'total_max_cap_rtd': 'Total Max Capacity (MW)'},
#              title='Total Max Capacity by major_part_id_new and plant_type',
#              hover_data={'resource_id': True, 'resource_name': True, 'plant_type': True, 'total_max_cap_rtd': ':.2f'},
#              )

# # Show the plot
# fig.show()


In [29]:
df_for_plotly

major_part_id_new plant_type  total_max_cap_rtd       resource_name  \
49                 AP       COAL              668.0  GNP DINGININ CFTPP   
50                 AP       COAL              668.0  GNP DINGININ CFTPP   
114               SMC       COAL              647.0          SUAL CFTPP   
113               SMC       COAL              647.0          SUAL CFTPP   
154               SMC       NATG              600.0         ILIJAN NGPP   
..                ...        ...                ...                 ...   
23                BSI       SOLR                1.0        BOSUNG SOLAR   
248              AHPC        HYD                0.9           AMLAN HEP   
409         EUROHYDRO        HYD                0.5          MARBEL HEP   
27                 AC        BAT                0.5      BATAAN BATTERY   
153              PEHC        HYD                0.3        CALIBATO HEP   

       resource_id  
49      01GNPD_U01  
50      01GNPD_U02  
114     01SUAL_G02  
113     01SUAL_G01  
154   03ILIJAN_G01  
..             ...  
23    01BOSUNG_G01  
248     06AMLA_G01  
409   14MARBEL_U01  
27   01BTSOLEN_BAT  
153   03CLBATO_G01  

[419 rows x 5 columns]

In [30]:
# Get the count of each 'major_part_id_new'
count_df = df.groupby('major_part_id_new').size().reset_index(name='count')

# Sort the DataFrame by count in descending order
sorted_major_part_df = count_df.sort_values(by='count', ascending=False)

# Group by 'major_part_id_new', 'plant_type', and 'resource_id' without aggregation
grouped_df = df.groupby(['major_part_id_new', 'plant_type', 'resource_id']).first().reset_index()

# Select only the desired columns
grouped_df = grouped_df[['major_part_id_new', 'plant_type', 'resource_id', 'resource_name', 'region_id', 'zone_id', 'max_cap_rtd']]

# Create a stacked bar chart using Plotly Express
fig = px.bar(grouped_df,
             x='major_part_id_new',
             color='plant_type',
             labels={'major_part_id_new': 'Holding Company', 'max_cap_rtd': 'Total Max Capacity (MW)'},
             title='Number of Participants by Holding Company and Plant Type',
             hover_data={'resource_id': True, 'plant_type': True},
             category_orders={'major_part_id_new': sorted_major_part_df['major_part_id_new']},
             )

# Show the plot
fig.show()


In [31]:
import plotly.express as px

# Assuming you have the original DataFrame called df

# Get the total max capacity for each 'major_part_id_new'
capacity_df = df.groupby('major_part_id_new')['max_cap_rtd'].sum().reset_index(name='total_max_capacity')

# Sort the DataFrame by total_max_capacity in descending order
sorted_capacity_df = capacity_df.sort_values(by='total_max_capacity', ascending=False)

# Group by 'major_part_id_new', 'plant_type', and 'resource_id' without aggregation
grouped_df = df.groupby(['major_part_id_new', 'plant_type', 'resource_id']).first().reset_index()

# Select only the desired columns
grouped_df = grouped_df[['major_part_id_new', 'plant_type', 'resource_id', 'resource_name', 'region_id', 'zone_id', 'max_cap_rtd']]

# Create a stacked bar chart using Plotly Express
fig = px.bar(grouped_df,
             x='major_part_id_new',
             y='max_cap_rtd',  # Use max_cap_rtd as the y-axis variable
             color='plant_type',
             labels={'major_part_id_new': 'Holding Company', 'max_cap_rtd': 'Total Max Capacity (MW)'},
             title='Total Max Capacity (in MW) by Holding Company and Plant Type',
             hover_data={'resource_id': True, 'plant_type': True},
             category_orders={'major_part_id_new': sorted_capacity_df['major_part_id_new']},
             )

# Show the plot
fig.show()


# Filtering

In [40]:
# Number of resource_id with at least 1 entry under rtd_quantity1 column
resource_count_with_quantity1 = df[df['rtd_quantity1'].notnull()]['resource_id'].nunique()

print(f"The number of resource_id with at least 1 entry under rtd_quantity1 column is: {resource_count_with_quantity1}")

The number of resource_id with at least 1 entry under rtd_quantity1 column is: 222


In [32]:
# Group by 'resource_id' and count the non-NaN entries in 'rtd_quantity1'
counts = df.groupby('resource_id')['rtd_quantity8'].count()

# Sort the counts in descending order and select the top 20
top_resources = counts.sort_values(ascending=False).head(20)
# Display the top 20 resource_id with non-NaN counts
for resource_id, count in top_resources.items():
    print(f"Resource ID: {resource_id}, Non-NaN Count: {count}")

Resource ID: 01SUAL_G01, Non-NaN Count: 288
Resource ID: 13SMC_U02, Non-NaN Count: 288
Resource ID: 10GNPK_U03, Non-NaN Count: 288
Resource ID: 10GNPK_U02, Non-NaN Count: 288
Resource ID: 03PAGBIL_G03, Non-NaN Count: 288
Resource ID: 01MSINLO_G01, Non-NaN Count: 288
Resource ID: 01MSINLO_G03, Non-NaN Count: 288
Resource ID: 13SMC_U01, Non-NaN Count: 288
Resource ID: 01MSINLO_G02, Non-NaN Count: 288
Resource ID: 01SUAL_G02, Non-NaN Count: 288
Resource ID: 01HYPGRN_G01, Non-NaN Count: 268
Resource ID: 11FDC_U02, Non-NaN Count: 120
Resource ID: 11FDC_U01, Non-NaN Count: 96
Resource ID: 11STEAG_U01, Non-NaN Count: 96
Resource ID: 08PDPP3_H, Non-NaN Count: 0
Resource ID: 06STNEGB_G01, Non-NaN Count: 0
Resource ID: 07BDPP_U03, Non-NaN Count: 0
Resource ID: 07BDPP_U02, Non-NaN Count: 0
Resource ID: 07BDPP_U01, Non-NaN Count: 0
Resource ID: 06VMC_G02, Non-NaN Count: 0


In [33]:
pd.set_option('display.max_columns', None)

# Filter by 'resource_id' and time range
filtered_df = df[(df['resource_id'] == '01SUAL_G01')]

# Select specific columns
selected_columns = ['time_interval', 'resource_id', 'resource_name', 'trading_id', 'major_part_id_new',
                     'plant_type', 'region_id', 'zone_id', 'max_cap_rtd',
                     'rtd_price1', 'rtd_quantity1', 'rtd_price2', 'rtd_quantity2',
                     'rtd_price3', 'rtd_quantity3', 'rtd_price4', 'rtd_quantity4',
                     'rtd_price5', 'rtd_quantity5', 'rtd_price6', 'rtd_quantity6',
                     'rtd_price7', 'rtd_quantity7', 'rtd_price8', 'rtd_quantity8',
                     'rtd_price9', 'rtd_quantity9', 'rtd_price10', 'rtd_quantity10',
                     'rtd_price11', 'rtd_quantity11', 'max_offer_energy_rtd_price',
                     'max_offer_energy_rtd', 'energy_dipc', 'energy_dipc_price',
                     'energy_rtx', 'energy_rtx_price', 'energy_dipc_final',
                     'energy_dipc_final_price', 'pricing_condition_final']

# Apply the selection to the filtered DataFrame
result_df = filtered_df[selected_columns]

# Sort by 'time_interval'
result_df = result_df.sort_values(by='time_interval')
result_df

time_interval resource_id resource_name trading_id  \
8080  2023-09-24 00:05:00  01SUAL_G01    SUAL CFTPP       SMEC   
8110  2023-09-24 00:10:00  01SUAL_G01    SUAL CFTPP       SMEC   
8140  2023-09-24 00:15:00  01SUAL_G01    SUAL CFTPP       SMEC   
8170  2023-09-24 00:20:00  01SUAL_G01    SUAL CFTPP       SMEC   
8200  2023-09-24 00:25:00  01SUAL_G01    SUAL CFTPP       SMEC   
...                   ...         ...           ...        ...   
16570 2023-09-24 23:40:00  01SUAL_G01    SUAL CFTPP       SMEC   
16600 2023-09-24 23:45:00  01SUAL_G01    SUAL CFTPP       SMEC   
16630 2023-09-24 23:50:00  01SUAL_G01    SUAL CFTPP       SMEC   
16660 2023-09-24 23:55:00  01SUAL_G01    SUAL CFTPP       SMEC   
16690 2023-09-25 00:00:00  01SUAL_G01    SUAL CFTPP       SMEC   

      major_part_id_new plant_type region_id zone_id  max_cap_rtd  rtd_price1  \
8080                SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
8110                SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
8140                SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
8170                SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
8200                SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
...                 ...        ...       ...     ...          ...         ...   
16570               SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
16600               SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
16630               SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
16660               SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   
16690               SMC       COAL      CLUZ  NLUZON        647.0    -10000.0   

       rtd_quantity1  rtd_price2  rtd_quantity2  rtd_price3  rtd_quantity3  \
8080             0.0    -10000.0          250.0     5200.08          300.0   
8110             0.0    -10000.0          250.0     5200.08          300.0   
8140             0.0    -10000.0          250.0     5200.08          300.0   
8170             0.0    -10000.0          250.0     5200.08          300.0   
8200             0.0    -10000.0          250.0     5200.08          300.0   
...              ...         ...            ...         ...            ...   
16570            0.0    -10000.0          250.0     5000.08          300.0   
16600            0.0    -10000.0          250.0     5000.08          300.0   
16630            0.0    -10000.0          250.0     5000.08          300.0   
16660            0.0    -10000.0          250.0     5000.08          300.0   
16690            0.0    -10000.0          250.0     5000.08          300.0   

       rtd_price4  rtd_quantity4  rtd_price5  rtd_quantity5  rtd_price6  \
8080      5700.08          350.0     6000.08          400.0     6500.08   
8110      5700.08          350.0     6000.08          400.0     6500.08   
8140      5700.08          350.0     6000.08          400.0     6500.08   
8170      5700.08          350.0     6000.08          400.0     6500.08   
8200      5700.08          350.0     6000.08          400.0     6500.08   
...           ...            ...         ...            ...         ...   
16570     6000.08          350.0     6100.08          400.0     6500.08   
16600     6000.08          350.0     6100.08          400.0     6500.08   
16630     6000.08          350.0     6100.08          400.0     6500.08   
16660     6000.08          350.0     6100.08          400.0     6500.08   
16690     6000.08          350.0     6100.08          400.0     6500.08   

       rtd_quantity6  rtd_price7  rtd_quantity7  rtd_price8  rtd_quantity8  \
8080           450.0     7000.08          500.0     7400.08          550.0   
8110           450.0     7000.08          500.0     7400.08          550.0   
8140           450.0     7000.08          500.0     7400.08          550.0   
8170           450.0     7000.08          500.0     7400.08          550.0   
8200           450

## [WIP] Sampling the Bidding Curve

In [34]:
import pandas as pd
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

# Assuming df is your DataFrame
# Replace 'time_interval', 'resource_id', and other column names as needed

# Set the specific Dispatching Unit (DU)
du_id = '07BDPP_U04'

# Filter the DataFrame for the specific DU
du_df = df[df['resource_id'] == du_id]

# Extract the maximum quantity (Qmax) from the 'max_cap_rtd' column
q_max = du_df['max_cap_rtd'].iloc[0]

In [35]:
# # Create a new DataFrame for bidding curves
# bidding_df = pd.DataFrame()

# # Loop through price-quantity pairs and append to bidding curves DataFrame
# for price_col, quantity_col, time_col in zip(price_columns, quantity_columns, [time_column]*11):
#     bidding_df = pd.concat([bidding_df, du_df[[time_col, price_col, quantity_col]].rename(columns={price_col: 'price', quantity_col: 'quantity'})])

In [36]:
# bidding_df

May mali sa for loop

for i in rtd_price1, rtd_quantity1 in range 1 to 11...

In [37]:
# Define the columns containing price-quantity pairs and timestamps
price_columns = [f'rtd_price{i}' for i in range(1, 12)]
quantity_columns = [f'rtd_quantity{i}' for i in range(1, 12)]
time_column = 'time_interval'

# Create a new DataFrame for bidding curves
bidding_df = pd.DataFrame()

# Loop through price-quantity pairs and append to bidding curves DataFrame
for i, (price_col, quantity_col, time_col) in enumerate(zip(price_columns, quantity_columns, [time_column]*11), start=1):
    # Extract relevant columns from du_df
    selected_columns = du_df[[time_col, price_col, quantity_col]].rename(columns={price_col: 'price', quantity_col: 'quantity'})

    # Add a new column 'M' with the value of 'i'
    selected_columns['M'] = i

    # Concatenate to the bidding_df DataFrame
    bidding_df = pd.concat([bidding_df, selected_columns])

# Reset index for the final DataFrame
bidding_df.reset_index(drop=True, inplace=True)
bidding_df

time_interval     price  quantity   M
0    2023-09-24 00:05:00  21227.75       0.0   1
1    2023-09-24 00:10:00  21227.75       0.0   1
2    2023-09-24 00:15:00  21227.75       0.0   1
3    2023-09-24 00:20:00  21227.75       0.0   1
4    2023-09-24 00:25:00  21227.75       0.0   1
...                  ...       ...       ...  ..
3163 2023-09-24 23:40:00       NaN       NaN  11
3164 2023-09-24 23:45:00       NaN       NaN  11
3165 2023-09-24 23:50:00       NaN       NaN  11
3166 2023-09-24 23:55:00       NaN       NaN  11
3167 2023-09-25 00:00:00       NaN       NaN  11

[3168 rows x 4 columns]

In [38]:
bidding_df_no_duplicates = bidding_df.drop_duplicates(subset=['price', 'quantity', 'M'])#.reset_index(drop=True)
bidding_df_no_duplicates = bidding_df_no_duplicates.sort_values(by='time_interval')
bidding_df_no_duplicates

time_interval     price  quantity   M
0    2023-09-24 00:05:00  21227.75       0.0   1
288  2023-09-24 00:05:00  21227.75       4.0   2
576  2023-09-24 00:05:00       NaN       NaN   3
864  2023-09-24 00:05:00       NaN       NaN   4
1152 2023-09-24 00:05:00       NaN       NaN   5
...                  ...       ...       ...  ..
2304 2023-09-24 00:05:00       NaN       NaN   9
2592 2023-09-24 00:05:00       NaN       NaN  10
2880 2023-09-24 00:05:00       NaN       NaN  11
108  2023-09-24 09:05:00  24000.00       0.0   1
396  2023-09-24 09:05:00  24000.00       4.0   2

[13 rows x 4 columns]